server/fleet/service.go

```diff
@@ -10,6 +10,14 @@ import (
 	"github.com/kolide/kit/version"
 )
 
+type Hijacker interface {
```

Comment:
To me, this seems very similar to defining something like 

```
// EnterpriseService contains methods that are only implemented in ee/...
type EnterpriseService interface {
}
```

In server/service/service.go, we could do something like 

```
type Service struct {
    // ...
    Enterprise fleet.EnterpriseService
}

func (s *Service) A() {
    // ...
    if s.Enterprise != nil {
        s.Enterprise.B()
    }
}
```

The advantage of this is that the non-enterprise `service.Service` struct would not need to provide default implementations of the enterprise only methods.

`HijackWith` could be replaced with just passing a non-nil `EnterpriseService` to the `Service` constructor.

Reply:
nice! I think this example is lacking context, as you will very much want a free implementation too.

An argument could be made that the free implementation could live in `A`:

```go
func (s *Service) A() {
    thing := freeDefault
    if s.Enterprise != nil {
        thing  = s.Enterprise.B()
    }
   // ... other stuff
}
```

but if you need to use `B` in more than one place, you'll probably end up with an abstraction:

```
func (s *Service) C() {
    thing := freeDefault
    if s.Enterprise != nil {
        thing  = s.Enterprise.B()
    }
}

func (s *Service) A() {
    thing := s.C()
   // ... other stuff
}
``` 

which, in concrete terms would mean having `ExampleMethod` and `Enterprise.ExampleMethod`, which sounds less mysterious than the hijacker, so I like it better. Do you have any thoughts?

Reply:
Both approach are fine with me, I think @michalnicp's approach is more straight forward. 

Perhaps another angle to look at this is that the current problem (IMHO) is with service composition, instead of calling services within services, ideally each service should be orthogonal and we should have different *workflows* for free and ee users:

*workflow (free)* : AFree => BFree => CFree
*workflow (ee)* : AFree => BFree => CEE